Library/Homebrew/cmd/upgrade.rb

```diff
@@ -171,4 +181,208 @@ def upgrade_formula(f)
       nil
     end
   end
+
+  def upgradable_dependents(kegs, formulae)
+    formulae_to_upgrade = Set.new
+    formulae_pinned = Set.new
+
+    formulae.each do |formula|
+      descendants = Set.new
+
+      dependents = kegs.select do |keg|
+        keg.runtime_dependencies
+           .any? { |d| d["full_name"] == formula.full_name }
+      end
+
+      next if dependents.empty?
+
+      dependent_formulae = dependents.map(&:to_formula)
+
+      dependent_formulae.each do |f|
+        next if formulae_to_upgrade.include?(f)
+        next if formulae_pinned.include?(f)
+
+        if f.outdated?(fetch_head: ARGV.fetch_head?)
+          if f.pinned?
+            formulae_pinned << f
+          else
+            formulae_to_upgrade << f
+          end
+        end
+
+        descendants << f
+      end
+
+      upgradable_descendants, pinned_descendants = upgradable_dependents(kegs, descendants)
+
+      formulae_to_upgrade.merge upgradable_descendants
+      formulae_pinned.merge pinned_descendants
+    end
+
+    [formulae_to_upgrade, formulae_pinned]
+  end
+
+  def broken_dependents(kegs, formulae)
+    formulae_to_reinstall = Set.new
+    formulae_pinned_and_outdated = Set.new
+
+    CacheStoreDatabase.use(:linkage) do |db|
+      formulae.each do |formula|
+        descendants = Set.new
+
+        dependents = kegs.select do |keg|
+          keg.runtime_dependencies
+             .any? { |d| d["full_name"] == formula.full_name }
+        end
+
+        next if dependents.empty?
+
+        dependents.each do |keg|
+          f = keg.to_formula
+
+          next if formulae_to_reinstall.include?(f)
+          next if formulae_pinned_and_outdated.include?(f)
+
+          checker = LinkageChecker.new(keg, cache_db: db)
+
+          if checker.broken_library_linkage?
+            if f.outdated?(fetch_head: ARGV.fetch_head?)
+              # Outdated formulae = pinned formulae (see function above)
+              formulae_pinned_and_outdated << f
+            else
+              formulae_to_reinstall << f
+            end
+          end
+
+          descendants << f
+        end
+
+        descendants_to_reinstall, descendants_pinned = broken_dependents(kegs, descendants)
+
+        formulae_to_reinstall.merge descendants_to_reinstall
+        formulae_pinned_and_outdated.merge descendants_pinned
+      end
+    end
+
+    [formulae_to_reinstall, formulae_pinned_and_outdated]
+  end
+
+  def check_dependents(formulae)
+    return if formulae.empty?
+
+    # First find all the outdated dependents.
+    kegs = Formula.installed
+                  .map(&:opt_or_installed_prefix_keg)
+                  .reject(&:nil?)
+                  .reject { |f| f.runtime_dependencies.to_a.empty? }
+
+    return if kegs.empty?
+
+    oh1 "Checking dependents for outdated formulae" if ARGV.verbose?
+    upgradable, pinned = upgradable_dependents(kegs, formulae).map(&:to_a)
+
+    upgradable.sort! do |a, b|
```

Comment:
Would be good to put these in a function as there's multiple identical versions.